src/cargo/sources/git/oxide.rs

```diff
@@ -0,0 +1,69 @@
+//! This module contains all code sporting `gitoxide` for operations on `git` repositories and it mirrors
+//! `utils` closely for now. One day it can be renamed into `utils` once `git2` isn't required anymore.
+
+use crate::util::{network, Progress};
+use crate::{CargoResult, Config};
+use git_repository as git;
+use std::sync::atomic::AtomicBool;
+use std::sync::Arc;
+use std::time::Duration;
+
+/// For the time being, `repo_path` makes it easy to instantiate a gitoxide repo just for fetching.
+/// In future this may change to be the gitoxide repository itself.
+pub fn with_retry_and_progress(
+    repo_path: &std::path::Path,
+    config: &Config,
+    cb: &mut (dyn FnMut(
+        &git::Repository,
+        &AtomicBool,
+        &mut git::progress::tree::Item,
+    ) -> CargoResult<()>
+              + Send),
+) -> CargoResult<()> {
+    let repo = git::open_opts(repo_path, {
+        let mut opts = git::open::Options::default();
+        // We need `git_binary` configuration as well for being able to see credential helpers
+        // that are configured with the `git` installation itself.
+        // However, this is slow on windows (~150ms) and most people won't need it as they use the
+        // standard index which won't ever need authentication.
+        // TODO: This is certainly something to make configurable, at the very least on windows.
+        //       Maybe it's also something that could be cached, all we need is the path to the configuration file
+        //       which usually doesn't change unless the installation changes. Maybe something keyed by the location of the
+        //       binary along with its fingerprint.
+        opts.permissions.config = git::permissions::Config::all();
+        opts
+    })?;
+
+    let progress_root: Arc<git::progress::tree::Root> = git::progress::tree::root::Options {
+        initial_capacity: 10,
+        message_buffer_capacity: 10,
+    }
+    .into();
+    let mut progress = progress_root.add_child("operation");
+
+    // For decent interrupts of long-running computations and removal of temp files we should handle interrupts, and this
+    // is an easy way to do that. We will remove them later.
+    // We intentionally swallow errors here as if for some reason we can't register handlers, `cargo` will just work like before and
+    // abort on signals.
+    let _deregister_signal_handlers_on_drop = git::interrupt::init_handler(|| {})
+        .ok()
+        .unwrap_or_default()
+        .auto_deregister();
```

Comment:
When writing this I also thought it might be a big step, but preferred to have it as basis of discussion.

Potentially long running operations in `gitoxide` always take a `should_interrupt: &AtomicBool` parameter to poll and be able to stop computations gracefully. Further, all operations that write files, like packs in this case, do so by using temporary files via `git-lock` and transitively `git-tempfile`. The latter registers tempfiles globally and offers an interrupt-capable handler which will remove tempfiles which aren't currently written. Both capabilities combined means that all temporary files can be cleaned up gracefully instead of remaining stale on disk.

`cargo` can and maybe _should_ rely on its built-in capabilities to clean up its repositories or refresh them if they seem broken, and probably that's the better choice here.

In other words, stale files in `cargo` was a solved problem, and there is no need to add another solution on top of it here.